### PR TITLE
Fix and Enhance StringConverter to handle more special characters

### PR DESCRIPTION
**Pull request by mizady**
## Description
- Fix and Enhance StringConverter to handle more special characters.
- Added test cases for StringConverter to validate special character handling during serialization and deserialization.

## Motivation and Context
- When serializing and/or de-serializing a file location such as I:\somefile.txt, the StringConverter was not correctly handling the case of a backslash.
- Fixes nanoFramework/Home#1562

## How Has This Been Tested?
- Added unit tests that verified the backslash, tab, backspace, and FeedForm special characters were not being handled correctly as escaped characters by the StringConverter.  All current tests in the nanoFramework.Json.Test project including the added tests are now passing.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by Ellerbach** [comment]
So, this is not properly working as explained in Discord. Here is the investigation: https://github.com/nanoframework/nanoFramework.Json/tree/pr361

And here are the elements:

* This will fail: [DataRow("I:\\nano\\rpath\\to", "I:\\nano\\rpath\\to")]  // More complex with combination of special characters and possible new line
* The string converter with the current logic only works for a json that is a single string
* In case the string is part of an existing obejct, the current logic will break when it's a string wontaining potential characters like in the previous example. Also, it will fail in case of a string of length 1

To solve this problem:

* As in the branch pr361 pointed above, do not use the escaped logic for objects, just comment out the code like in that branch
* For single string, then, the logic should be used

**Comment by Ellerbach** [comment]
And the case already exists:

* https://github.com/nanoframework/nanoFramework.Json/blob/f31157a19724bec6bbde21ae1e1fb2408a7a445b/nanoFramework.Json/JsonConvert.cs?plain=1

So, in short, it's about using the function I've commented for that case and calling it in the if. Otherwise, we should not use it.

**Comment by mizady** [comment]
I noticed that the discord thread was not actually linked here. [Discord thread](https://discord.com/channels/478725473862549535/1314972636857499681)

RFC8259 - https://datatracker.ietf.org/doc/html/rfc8259
ECMA-404 - https://ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf

Seems like we are handling all the escaped characters correctly now from what I can tell based on testing.

**Comment by Ellerbach** [comment]
Are the `\uXXXX`still properly handled with the new implementation?

**Comment by mizady** [comment]
> Are the `\uXXXX`still properly handled with the new implementation?

The original unicode processing part of he code is still used for unicode characters.  The existing tests that test unicode properly serializing and deserializing have all passed.

File: nanoFramework.Json/JsonConvert.cs
Comment (josesimoes) [inline nanoFramework.Json/JsonConvert.cs:1095]: Why all these are being commented out? If they aren't needed better delete them...

Comment (mizady) [reply nanoFramework.Json/JsonConvert.cs:1095]: To be honest, bad habit I picked up from school (leaving previous code since we did not use change tracking), i will fix that and commit.

Comment (mizady) [reply nanoFramework.Json/JsonConvert.cs:1095]: also while going over it again, i found yet another situation where it is not de-serializing correctly, going to add more tests to more thoroughly cover the characters instead of just checking the " and \ characters in the tests.

Comment (mizady) [reply nanoFramework.Json/JsonConvert.cs]: I have committed the updated/added tests and the modified logic in StringConverter and JsonConvert.  There are still two tests not correctly passing (one of which I added).  I am not sure if I am incorrect about the test values, as the same problem seems to exist in the test SerializeCosmosDbObject_02 which is failing for removed quotes as well, though the test SerializeCosmosDbObject_01 which uses almost the same string, seems to work fine.  I think another set of eyes may be necessary.

Comment (mizady) [reply nanoFramework.Json/JsonConvert.cs]: alright final code has been committed that replaces the quotes that should not be removed during de-serialization of strings that are part of complex objects which also contain an escaped quote at the start and end of the string. all tests should now be passing.

**Review by Ellerbach — COMMENTED** [review]
Before approving, can you add a test with a complex object, something that is already existing with a string that contains escaped chars so we are sure we cover properly that case as well?

For the rest, looks all goos! Thanks a lot for the efforts getting there.

**Comment by mizady** [comment]
The test labeled "Can_serialize_and_deserialize_object_containing_string_with_escaped_characters" located in the JsonSerializerOptionsTests has a data row for the case of escaped quotes at the beginning and end.  Also in the same test class is "Can_serialize_and_deserialize_string_with_escaped_characters" which has all the same data rows for a string that is not part of a complex object.  So unless I missed something (which of course is possible) then there should be tests handling all of these cases.

**Comment by mizady** [comment]
I've added 18 test cases to each of the two tests for simple string and complex object containing a string serialization and deserialization. this is now the list of tested combinations for both. Totaling 68 tests between the two.

[DataRow("a")]  // Single character
[DataRow("1")]  // Single numeric character
[DataRow("\t")]  // Single Tab character
[DataRow("Testing / solidus")] // Forward slash in string
[DataRow("Testing  solidus")] // Double space in string
[DataRow("Some string with \" that needs escaping")] // String containing a quote
[DataRow("Quotes in a \"string\".")] // String with escaped quotes
[DataRow("Escaped last character \n")] // Newline as the last character
[DataRow("I:\\Nano\\rApp\\app.pe")] // Backslash in string
[DataRow("Tab \t in a string \t")] // Tab character in multiple places
[DataRow("Newline \n in a string \n")] // Newline character in multiple places
[DataRow("LineFeed \f in a string \f")] // Line feed character in multiple places
[DataRow("CarriageReturn \r in a string \r")] // Carriage return character in multiple places
[DataRow("Backspace \b in a string \b")] // Backspace character in multiple places
[DataRow("TestString")] // Simple string with no special characters
[DataRow("\"TestString\"")] // String wrapped in quotes
[DataRow("\u0041")] // Unicode character (A)
[DataRow("\u2764")] // Unicode character (❤)
[DataRow("\x1B")] // Escape character (ASCII 27)
[DataRow("\x7F")] // Delete character (ASCII 127)
[DataRow("\0")] // Null character
[DataRow("")] // Empty string
[DataRow("Line 1\nLine 2\nLine 3")] // Multi-line string
[DataRow("Curly braces: { }")] // JSON-like curly braces
[DataRow("Square brackets: [ ]")] // JSON-like square brackets
[DataRow("Colon and comma: : ,")] // Colon and comma
[DataRow("Special symbols: @#$%^&*()_+~")] // Special symbols
[DataRow("English 中文 Español العربية हिंदी")] // Mixed language text
[DataRow("{\"key\": \"value\"}")] // JSON-like string
[DataRow("\"[{\"inner\":\"value\"}]\"")] // Serialized JSON-like string
[DataRow("{\"name\":\"John\",\"age\":30}")] // Serialized JSON
[DataRow("Invalid escape: \\q")] // Invalid escape sequence
[DataRow("https://example.com/api?query=escaped%20characters")] // URL
[DataRow("Unicode \u2764, Newline \n, Tab \t, Backslash \\")] // Combination of cases

**Review by Ellerbach — APPROVED** [review]
Great, thanks a lot for the fix!

**Comment by Ellerbach** [comment]
> I've added 18 test cases to each of the two tests for simple string and complex object containing a string serialization and deserialization

Amazing! Thanks a lot!

**Comment by Ellerbach** [comment]
@josesimoes we do have 2 of the actions not running, any idea why before I force the merge?

**Comment by josesimoes** [comment]
> @josesimoes we do have 2 of the actions not running, any idea why before I force the merge?

They were waiting for a maintainer approval to run. (github is getting heavier and heavier on security 😉 )

**Pull request merged**
